File: src/main/java/com/example/finsentinel/service/trading/uta/AlpacaBroker.java
package com.example.finsentinel.service.trading.uta;

import com.example.finsentinel.service.trading.engine.AccountInfo;
import com.example.finsentinel.service.trading.engine.AlpacaTradingEngine;
import com.example.finsentinel.service.trading.engine.MarketClock;
import com.example.finsentinel.service.trading.engine.OrderRequest;
import com.example.finsentinel.service.trading.engine.OrderResult;
import com.example.finsentinel.service.trading.engine.PositionInfo;

import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * IBroker adapter for Alpaca US equities trading.
 * Supports STOCK security type only.
 * Not a Spring bean — created by BrokerRegistry.
 */
public class AlpacaBroker implements IBroker {

    private final AlpacaTradingEngine engine;

    public AlpacaBroker(AlpacaTradingEngine engine) {
        this.engine = Objects.requireNonNull(engine, "AlpacaTradingEngine must not be null");
    }

    @Override
    public String brokerId() {
        return "alpaca";
    }

    @Override
    public String displayName() {
        return "Alpaca (US Equities)";
    }

    @Override
    public Set<SecurityType> supportedSecurityTypes() {
        return EnumSet.of(SecurityType.STOCK);
    }

    @Override
    public Set<BrokerCapability> capabilities() {
        return EnumSet.of(
                BrokerCapability.SPOT_TRADING,
                BrokerCapability.MARKET_DATA,
                BrokerCapability.ORDER_MANAGEMENT,
                BrokerCapability.SHORT_SELLING);
    }

    @Override
    public OrderResult placeOrder(Contract contract, OrderRequest request) {
        var nativeRequest = new OrderRequest(
                contract.toEngineSymbol(), request.side(), request.type(),
                request.qty(), request.notional(), request.price(),
                request.stopPrice(), request.timeInForce(), request.reduceOnly());
        return engine.placeOrder(nativeRequest);
    }

    @Override
    public List<PositionInfo> getPositions() {
        return engine.getPositions();
    }

    @Override
    public List<OrderResult> getOrders() {
        return engine.getOrders();
    }

    @Override
    public List<OrderResult> syncOrders() {
        return engine.syncOrders();
    }

    @Override
    public AccountInfo getAccount() {
        return engine.getAccount();
    }

    @Override
    public boolean cancelOrder(String orderId) {
        return engine.cancelOrder(orderId);
    }

    @Override
    public MarketClock getMarketClock() {
        return engine.getMarketClock();
    }
}

File: apps/api/src/trading/brokers/alpaca.broker.ts
import { SecurityType, BrokerCapability, Contract } from '@finsentinel/shared';
import type { IBroker } from '../interfaces/broker';
import type { AlpacaTradingEngine } from '../engines/alpaca-trading.engine';
import type {
  OrderRequest,
  OrderResult,
  PositionInfo,
  AccountInfo,
  MarketClock,
} from '../interfaces/types';

/**
 * Alpaca broker capabilities — spot trading only (US equities).
 */
const ALPACA_CAPABILITIES: Set<BrokerCapability> = new Set([BrokerCapability.SPOT_TRADING]);

/**
 * AlpacaBroker — wraps AlpacaTradingEngine with Contract-aware symbol conversion.
 *
 * NOT an Injectable. Created by BrokerRegistry at runtime.
 *
 * Alpaca broker adapter:
 * - Supports STOCK security type only
 * - Converts contract to engine-native symbol via Contract.toEngineSymbol()
 * - Delegates all operations to the underlying AlpacaTradingEngine
 */
export class AlpacaBroker implements IBroker {
  constructor(private readonly _engine: AlpacaTradingEngine) {}

  // -- Identity ---------------------------------------------------------------

  brokerId(): string {
    return 'alpaca';
  }

  displayName(): string {
    return 'Alpaca (US Equities)';
  }

  // -- Capability checks ------------------------------------------------------

  supportedSecurityTypes(): Set<SecurityType> {
    return new Set([SecurityType.STOCK]);
  }

  capabilities(): Set<BrokerCapability> {
    return ALPACA_CAPABILITIES;
  }

  canHandle(contract: Contract): boolean {
    return contract.secType === SecurityType.STOCK;
  }

  // -- Order execution --------------------------------------------------------

  async placeOrder(contract: Contract, request: OrderRequest): Promise<OrderResult> {
    const adapted: OrderRequest = {
      ...request,
      symbol: contract.toEngineSymbol(),
    };

    return this._engine.placeOrder(adapted);
  }

  // -- Delegated operations ---------------------------------------------------

  async getPositions(): Promise<PositionInfo[]> {
    return this._engine.getPositions();
  }

  async getOrders(): Promise<OrderResult[]> {
    return this._engine.getOrders();
  }

  async getAccount(): Promise<AccountInfo> {
    return this._engine.getAccount();
  }

  async cancelOrder(orderId: string): Promise<boolean> {
    return this._engine.cancelOrder(orderId);
  }

  async syncOrders(): Promise<OrderResult[]> {
    return this._engine.syncOrders!();
  }

  async getMarketClock(): Promise<MarketClock> {
    return this._engine.getMarketClock!();
  }

  async searchContracts(_query: string): Promise<Contract[]> {
    return [];
  }

  // -- Engine access ----------------------------------------------------------

  /** Expose the underlying engine for direct access when needed. */
  engine(): AlpacaTradingEngine {
    return this._engine;
  }
}
